Implement Math.cbrt(), Math.log{2,10}(), Math.trunc()

fatcerberus opened the pull request:
This pull adds support for ES6 `Math.cbrt()`, which calculates the cubic root of a number, as well as `Math.log2()` and `Math.log10()` (base 2 and base 10 logarithm).  The C99 standard functions are used if available (i.e. for C99/C++11, MSVC 2013+), otherwise a workaround is used instead.

Checklist:
- [x] Implement support for `Math.cbrt()`
- [x] Implement support for `Math.log2()`, `Math.log10()`
- [x] Implement support for `Math.trunc()`
- [x] Testcases
- [x] Releases entry

fatcerberus commented:
MSVC 2012 seems to be missing `cbrt()`.  Not surprising, C99 support in MSVC was abysmal prior to v2015.

svaarala commented:
The easiest solution for platform_fillins.h would be to define `DUK_CBRT` narrowly when it is known to exist:
- C99 or above
- C++11 or above
- MSVC above a certain version

Otherwise leave it undefined.

fatcerberus commented:
Okay, MSVC builds (AppVeyor) look like they're passing now.

fatcerberus commented:
@svaarala This is ready to go, just awaiting final test results and review.

fatcerberus commented:
I added `log2()` and `log10()` to the pull, since they are generally useful (unlike, say, `Math.clz32()`) and have trivial workarounds (log division).

fatcerberus commented:
The log2 and log10 tests are failing, I'll have to look into that tonight.

fatcerberus commented:
Added `Math.trunc()`.  I think I'm done now :)

svaarala commented on tests/ecmascript/test-bi-math-cbrt.js:
```diff
@@ -0,0 +1,47 @@
+
+    printExact(Math.cbrt(0));
+    printExact(Math.cbrt(-0));
+    printExact(Math.cbrt(Infinity));
```
This is fine but generally testcases use `1/0`, `-1/0`, and `0/0` for +Inf, -Inf, and NaN. Not important as such, just avoids a dependency on the global binding.

fatcerberus replied:
Hm, I guess that makes sense for purity since Duktape allows you to remove bindings at compile time.  I wrote the above under the assumption that the global constants would always be available (they are non-writable and non-configurable, so this is a reasonable assumption in most environments), but for Duktape that may not always hold.

svaarala replied:
Testcases don't need to support those cases though. This was more a comment on consistency. Functionally it's fine to reference Infinity etc.

svaarala replied:
It's not consistently applied but I've tried to start using `void 0` for `undefined` too; as an idiom it generates better code and doesn't rely on the weird global binding. Again, more of a style choice than anything.

fatcerberus replied:
Alright, I'll change the tests before the merge and save you some trouble :)

svaarala commented on tests/ecmascript/test-bi-math-log2.js:
```diff
@@ -0,0 +1,44 @@
+    printExact(Math.log2(2));
+    printExact(Math.log2(1024));
+    printExact(Math.log2(2.73121871170759e+244));
+    printRounded6(Math.log2(1000));
```
One interesting test for log2 would be exactness for inputs like 2^n where n is in a reasonable range (at least 0 to 53, i.e. the exact integer range).

I know no guarantees are in the specification but for practical purposes this is probably quite interesting. The guarantee holds for V8 at least for the entire IEEE double exponent range.

svaarala replied:
The guarantee won't hold for the C99 `log2()` replacement but it'd still be nice to document the current behavior when the C99 binding is available. Perhaps in a separate testcase, `log2-2ton.js` or something.

svaarala replied:
Here's an example test which checks for both (prints nothing when everything checks out):

```
> expect = 0; x = 1; while (x !== 1/0) { if (Math.log2(x) != expect) { console.log(x, Math.log2(x)); } x *= 2; expect++; }
1023
> expect = 0; x = 1; while (x !== 0) { if (Math.log2(x) != expect) { console.log(x, Math.log2(x)); } x /= 2; expect--; }
-1074
``'

svaarala replied:
(The test prints out the last number because of the way Node.js console evals input; it's not actually console.log()ed by the loop.)

fatcerberus commented on src-input/duk_bi_math.c:
```diff
@@ -90,6 +90,54 @@ DUK_LOCAL double duk__fmax_fixed(double x, double y) {
+#if defined(DUK_LOG2)
+	return DUK_LOG2(x);
+#else
+	return DUK_LOG(x) / DUK_LOG(2.0);
```
I considered that too, but then I remembered reading that most compilers will optimize e.g. `log(2.0)` to a constant during constant folding anyway.  So I kept it this way for clarity.  Your call, though.

svaarala replied:
That's quite likely, but they can't do so when `log()` is provided by a custom library which the compiler doesn't know about. But I can make the change separately.

fatcerberus commented on src-input/duk_bi_math.c:
```diff
@@ -90,6 +90,54 @@ DUK_LOCAL double duk__fmax_fixed(double x, double y) {
+	 * identical, but that's OK: ES6 says it's implementation-dependent.
+	 */
+
+	duk_small_int_t c = (duk_small_int_t) DUK_FPCLASSIFY(x);
```
Alright, those are easy fixes :)

fatcerberus commented:
I fixed the tests to use computed Infinity and NaN and renamed the helper functions.  Let me know if there are any other changes you want to make in this pull.

svaarala commented:
I'd be interested in the Math.log2() exact 2^n test coverage - but other than that just squashing the fixes. :)

fatcerberus commented:
`duk__cbrt()` etc. helpers should probably be hidden behind an `#ifdef DUK_USE_ES6` but I didn't do that here because it was awkward with the lookup table for the onearg/twoarg helpers.

svaarala commented:
Hmm, right.. They'll then be pulled in even for low memory builds. Let's see how that affects footprint once the tests clear. It's probably going to be ~200-300 bytes if I had to guess, maybe enough to fix in a follow-up (not a blocker here IMO).

That array would need to be structured so that it contains the low memory (ES5) bindings first, followed by ES6 binding which will be absent. Awkward, as you said.

fatcerberus commented:
By the way it's impossible to run individual testcases directly from the command-line using duk now - the include file doesn't get pulled in.  That was an inconvenient surprise ;-)

svaarala commented:
Yep, there's `util/prep_test.py` which does the preparation but it's not really friendly to use. I'll try to merge in `tools/runtest.py` soon so there's something easier to use.

svaarala commented:
Hmm, you should probably be able to use `tools/runtest.py` just by copying it off the branch if you wanted to try it right now. But up to you naturally, I'll try to get it merged and test it on Windows myself too.

fatcerberus commented:
Thanks, I forgot about #1094, I still have to try that out.  After I prep this pull for merge I'll go look at that.

fatcerberus commented:
I added a test for `Math.log2(2 ** n)` for exponents 0-53, like you suggested.  Ran the test suite manually (it's way quicker on my i7-6700K than waiting for Travis), and everything passes, so looks like we're good. :)

svaarala commented:
Ok, I'll wait for Travis to see the footprint results and then merge.

fatcerberus commented:
I could have sworn I fixed those tabs in the .js file...

I recently had to wipe Notepad++ settings as it was locking up on startup, so I lost all my space/tab settings.  I keep forgetting to change them back.

svaarala commented:
Impact for minsize build on x86 is 200 bytes, roughly as expected.

fatcerberus commented:
Okay, code policy is good now, this is ready to be merged.

svaarala commented:
Great, merged :)
